Sources/OpenCombine/Publishers/Publishers.Zip.swift

```diff
@@ -0,0 +1,756 @@
+    {
+        switch completion {
+        case .failure:
+            downstream.receive(completion: completion)
+            let subscriptionsToCancel: [Subscription] = lock.do {
+                child.state = .failed
+                return upstreamSubscriptions
+            }
+            subscriptionsToCancel.forEach { $0.cancel() }
+        case .finished:
+            let shouldProcessQueue: Bool = lock.do {
+                child.state = .finished
+                queuedWork.append(.receivedFinishedFromUpstream)
+                if !queueIsBeingProcessed {
+                    queueIsBeingProcessed = true
+                    return true
+                }
+                return false
+            }
+
+            if shouldProcessQueue {
+                processQueue()
+            }
+        }
+    }
+
+    private func maybeDequeueValue() -> Downstream.Input? {
+        return hasCompleteValueAvailable ? dequeueValue() : nil
+    }
+
+    private func sendSubscriptionDownstream() {
+        downstream.receive(subscription: self)
+    }
+
+    private var hasCompleteValueAvailable: Bool {
+        return upstreamSubscriptions.allSatisfy { $0.hasValue }
+    }
+
+    private var areMoreValuesPossible: Bool {
+        // More values are possible if all children are (active || have surplus)
+        return upstreamSubscriptions
+            .allSatisfy { $0.state == .active || $0.hasValue }
+    }
+
+    private enum QueueAction {
+        case stopProcessing
+        case noAction
+        case sendFinishDownstream
+        case sendValueDownstream(_ value: Downstream.Input)
+        case sendRequestUpstream(_ demand: Subscribers.Demand)
+    }
+
+    private enum QueuedWork {
+        case receivedValueFromUpstream(value: Downstream.Input)
+        case receivedFinishedFromUpstream
+        case receivedRequestFromDownstream(demand: Subscribers.Demand)
+    }
+
+    private func lockedActionToTake() -> QueueAction {
+        guard let work = queuedWork.popFirst() else { return .stopProcessing }
+        switch work {
+        case .receivedValueFromUpstream(let value):
+            // TODO: Fix the implementation of Demand. I think it currently is too
+            // strict given that the documentation says:
+            //      any operation that would result in a negative value is
+            //      clamped to .max(0).
+            //It doesn't say anything about fatalErrors
+            if downstreamDemand != .none {
+                downstreamDemand -= 1
+            }
+            return .sendValueDownstream(value)
+        case .receivedFinishedFromUpstream:
+            return (processingValueForChild != nil || areMoreValuesPossible) ?
+                .noAction : .sendFinishDownstream
+        case .receivedRequestFromDownstream(let demand):
+            return .sendRequestUpstream(demand)
+        }
+    }
+
+    @discardableResult
+    private func processQueue() -> Subscribers.Demand? {
```

Comment:
I had a closer look and not sure what you mean, the lock used here is already `UnfairRecursiveLock`, is there anything else that you'd like to be changed?

Reply:
Basically:

https://github.com/OpenCombine/OpenCombine/blob/48d22f455a8edb22d31f5806a98bfeffd1c0cdb8/Sources/OpenCombine/Publishers/Publishers.Zip.swift#L440-L442

The whole queue thing was introduced to avoid calling any client code with the (non-recursive) lock held. But the same can be achieved by just using the recursive lock. Since we now use it, the point of having to deal with the queue is lost. What I suggest is let's get rid of the whole queue thing. It's very hard to understand.

Reply:
Thanks, the queue code has been removed.